src/sync/sync_manager.cpp

```diff
@@ -295,48 +298,64 @@ bool SyncManager::perform_metadata_update(std::function<void(const SyncMetadataM
     return true;
 }
 
-std::shared_ptr<SyncUser> SyncManager::get_user(const SyncUserIdentifier& identifier,
+std::shared_ptr<SyncUser> SyncManager::get_user(const std::string& user_id,
                                                 std::string refresh_token,
-                                                std::string access_token)
+                                                std::string access_token,
+                                                const std::string provider_type)
 {
-    std::lock_guard<std::mutex> lock(m_user_mutex);
-    auto it = m_users.find(identifier);
+    std::lock_guard<std::recursive_mutex> lock(m_user_mutex);
+    auto it = std::find_if(m_users.begin(),
+                           m_users.end(),
+                           [user_id, provider_type](const auto& user) {
+        return user->identity() == user_id && user->provider_type() == provider_type;
+    });
     if (it == m_users.end()) {
         // No existing user.
         auto new_user = std::make_shared<SyncUser>(std::move(refresh_token),
-                                                   identifier.user_id,
-                                                   identifier.auth_server_url,
-                                                   std::move(access_token));
-        m_users.insert({ identifier, new_user });
+                                                   user_id,
+                                                   provider_type,
+                                                   std::move(access_token),
+                                                   SyncUser::State::Active);
+        m_users.emplace(m_users.begin(), new_user);
         return new_user;
     } else {
-        auto user = it->second;
+        auto user = *it;
         if (user->state() == SyncUser::State::Error) {
             return nullptr;
         }
         user->update_refresh_token(std::move(refresh_token));
         user->update_access_token(std::move(access_token));
+
+        if (user->state() == SyncUser::State::LoggedOut) {
+            user->set_state(SyncUser::State::Active);
+        }
         return user;
     }
 }
 
-std::vector<std::shared_ptr<SyncUser>> SyncManager::all_logged_in_users() const
+std::vector<std::shared_ptr<SyncUser>> SyncManager::all_users()
```

Comment:
This function can be simplified quite a bit now:

```
    std::lock_guard<std::mutex> lock(m_user_mutex);
    m_users.erase(std::remove_if(m_users.begin(), m_users.end(), [](auto& user) {
        return user->state() == SyncUser::State::Error;
    }), m_users.end());
    return m_users;
```

Most of what it was doing was related to that `m_users` used to be a different type from what this function returns, but with that no longer being the case we can just return `m_users` and let the copy constructor do the copying.